third_party/move/move-vm/types/src/value_serde.rs

```diff
@@ -0,0 +1,411 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::values::{
+    Container, DeserializationSeed, SerializationReadyValue, SizedID, Struct, Value, ValueImpl,
+};
+use move_binary_format::errors::{PartialVMError, PartialVMResult};
+use move_core_types::{
+    value::{IdentifierMappingKind, MoveTypeLayout},
+    vm_status::StatusCode,
+};
+use serde::{
+    de::{DeserializeSeed, Error as DeError},
+    ser::Error as SerError,
+    Deserialize, Deserializer, Serialize, Serializer,
+};
+use std::{collections::HashSet, hash::Hash, str::FromStr};
+
+pub trait CustomDeserialize {
+    fn custom_deserialize<'d, D: Deserializer<'d>>(
+        &self,
+        deserializer: D,
+        tag: &IdentifierMappingKind,
+        layout: &MoveTypeLayout,
+    ) -> Result<Value, D::Error>;
+}
+
+pub trait CustomSerialize {
+    fn custom_serialize<S: Serializer>(
+        &self,
+        serializer: S,
+        tag: &IdentifierMappingKind,
+        layout: &MoveTypeLayout,
+        sized_id: SizedID,
+    ) -> Result<S::Ok, S::Error>;
+}
+
+// FIXME: consolidate
+fn is_string_layout(layout: &MoveTypeLayout) -> bool {
+    use MoveTypeLayout::*;
+    if let Struct(move_struct) = layout {
+        if let [Vector(elem)] = move_struct.fields().iter().as_slice() {
+            if let U8 = elem.as_ref() {
+                return true;
+            }
+        }
+    }
+    false
+}
+
+fn is_derived_string_struct_layout(layout: &MoveTypeLayout) -> bool {
+    use MoveTypeLayout::*;
+    if let Struct(move_struct) = layout {
+        if let [value_field, Vector(padding_elem)] = move_struct.fields().iter().as_slice() {
+            if is_string_layout(value_field) {
+                if let U8 = padding_elem.as_ref() {
+                    return true;
+                }
+            }
+        }
+    }
+    false
+}
+
+fn size_u32_as_uleb128(mut value: usize) -> usize {
+    let mut len = 1;
+    while value >= 0x80 {
+        // 7 (lowest) bits of data get written in a single byte.
+        len += 1;
+        value >>= 7;
+    }
+    len
+}
+
+fn u64_to_fixed_size_utf8_bytes(value: u64, length: usize) -> Option<Vec<u8>> {
+    let result = format!("{:0>width$}", value, width = length)
+        .to_string()
+        .into_bytes();
+    if result.len() != length {
+        None
+    } else {
+        Some(result)
+    }
+}
+
+fn bcs_size_of_byte_array(length: usize) -> usize {
+    size_u32_as_uleb128(length) + length
+}
+
+fn bytes_and_width_to_derived_string_struct(bytes: Vec<u8>, width: usize) -> Option<Value> {
```

Comment:
can we remove this, and some of the other functions from the other place? (at least those that we can import and don't directly use there)

Reply:
Problem before was that we are under third-party, and the functions you have previously defined are in aptos-types. I ended up moving these things to third-party `delayed_values` crate, no duplication anymore, but we might need to think on where all of these belong :)